src/server/chunk_server/chunk_server_impl.cc

```diff
@@ -170,11 +179,37 @@ bool ChunkServerImpl::ReportToMaster() {
   return successful_report > 0;
 }
 
+void ChunkServerImpl::StartReportToMaster() {
+  chunk_reporting_thread_ = new std::thread([&]() {
+    // Periodically run ReportToMaster until instructed to terminate
+    while (!reporting_thread_terminated_.load()) {
+      if (!this->ReportToMaster()) {
```

Comment:
Its thread safe but you can still confirm. Leveldb handles the safety. From leveldb: "Within a single process, the same leveldb::DB object may be safely shared by multiple concurrent threads. I.e., **different threads may write into or fetch iterators or call Get on the same database without any external synchronization** (the leveldb implementation will automatically do the required synchronization)"

Reply:
I think GetAllFileChunkMetadata is thread-safe for the reason that Michael mentioned and furthermore, because it owns its own iterator so not shared with other threads. 

The following sentence after Micheal's quote says "However other objects (like Iterator and WriteBatch) may require external synchronization. If two threads share such an object, they must protect access to it using their own locking protocol. More details are available in the public header files." It mentioned Iterator above, and my understanding is that as long as we are not sharing leveldb iterator's among threads we are safe.

Reply:
Yes we aren't sharing the iterator between threads. Each thread creates its own iterator. So that is safe.